lexical ordering over nats

zstone1 opened the pull request:
lexical ordering over nats, uses more axioms than mathcomp's order, so it makes sense to put it in classical. This topology is mostly useful for finitely branching trees, where the product and this order topology agree. Such proofs are coming next. 

This PR slightly depends on the order topology PR for some definitions in `set_interval`, so review that one first. 
<!-- if this PR fixes an issue, use "fixes #XYZ" -->

<!-- you may also explain what remains to do if the fix is incomplete -->

##### Checklist

- [ ] added corresponding entries in `CHANGELOG_UNRELEASED.md`

<!-- rebasing often messes with CHANGELOG_UNRELEASED.md -->
<!-- consider using a temporary CHANGELOG_PR1234.md instead -->
<!-- only append to minimize problems when merging/rebasing -->
<!-- consider the use of `changelog/changes.sh` from
     https://github.com/math-comp/tools to generate the changelog -->

- [ ] added corresponding documentation in the headers

Reference: [How to document](https://github.com/math-comp/math-comp/wiki/How-to-document)

<!-- Cross-out the above items using ~crossed out item~ when irrelevant -->

##### Reminder to reviewers

- Read [this Checklist](https://github.com/math-comp/math-comp/wiki/Checklist-for-creating-and-review-PRs)
- Put a milestone if possible
- Check labels

affeldt-aist commented:
About the naming:
- `share_prefix_leq` is maybe better than `share_prefixS`
- `share_prefixC` -> `share_prefix_sym` and `first_diffC `-> `first_diff_sym` felt a bit more natural to me
- what about `same_prefix` rather than `share_prefix`? (one less character and that avoids to mistype `shared_prefix`)

affeldt-aist commented:
In the last commit, I edited the script to avoid unfolding `Order.lt` and `Order.le`.
For example, `order.v` provides `leEnat` to switch from `Order.le` to `leq`,
so I introduced a similar lemma for `big_lexi_order` (`leEbig_lexi_order`).

affeldt-aist reviewed (approved):
I am a bit bothered by the naming. There is a section `big_lexi_order` but the identifier `big_lexi_order` is defined outside of it. Maybe it should be defined before the section since it does not depend on it and it could also be named with a more generic name to reflect the fact that it is about dependent functions. Then it becomes possible to rename `big_lexi_ord` into `big_lexi_order` which would better reflect that its type is `rel` (`ord` is a bit too reminiscent of "ordinal").

`lexi_bigprod_prefix_lt`, `lexi_bigprod_prefix_gt`, `lexi_bigprod_between` have the substring `lexi_bigprod` but the latter does not appear explicitly in the statement.
I am wondering whether there isn't a better name using identifiers that appear
in the statement but I have no proposal right now.

zstone1 commented:
Yeah, I'm looking back and realizing the naming problems stem from `lexi_bigprod`. Just pushed a version where 
1. `lexi_bigprod` -> `big_lexi_le`
2. Renamed the section to `big_lexi_le`, which is much more honest
3. `big_lexi_order` is now defined at the top of the file. I think this name makes more sense now, so that the `big_lexi_order` type has `big_lexi_le` as its order.
4. `big_lexi_ord` is removed. It doesn't need to exist, and it certainly shouldn't have escaped that section. And your points about ordinals is particularly relevant here. This all generalizes easily to work over any well-ordered set, not just `nat`. 
5. since `lexi_bigprod` doesn't exist, I renamed all those lemmas to `big_lexi_order_prefix_gt`, etc

Things are a bit more internally consistent with these names.

affeldt-aist commented:
This PR (now merged) is causing the CI of Coq to fail: https://github.com/math-comp/analysis/issues/1333
